File: HeartLayout/app/src/main/java/com/example/clevo/heartlayout/MainActivity.java
package com.example.clevo.heartlayout;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.example.clevo.heartlayout.view.HeartView;

public class MainActivity extends AppCompatActivity {

    HeartView main_heartview;
    Button main_button;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        main_heartview= (HeartView) findViewById(R.id.main_heartview);
        main_button= (Button) findViewById(R.id.main_button);
        main_button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                main_heartview.addHeart();
            }
        });
    }
}

File: LogDemo/app/src/main/java/com/renyu/logdemo/MainActivity.kt
package com.renyu.logdemo

import android.os.Bundle
import android.widget.Button
import android.widget.ScrollView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    private val scroll by lazy {
        findViewById<ScrollView>(R.id.scroll)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.stop).setOnClickListener {
            LogcatExecutor.stopOutput()
        }

        LogcatExecutor.startOutput(object : LogcatExecutor.CallBack {
            override fun onLogOutput(log: String) {
                runOnUiThread {
                    findViewById<TextView>(R.id.content).text = log
                    scroll.post {
                        scroll.fullScroll(ScrollView.FOCUS_DOWN)
                    }
                }
            }
        })
    }
}